src/file_pipe_log/pipe.rs

```diff
@@ -308,12 +433,52 @@ impl<F: FileSystem> SinglePipe<F> {
                     continue;
                 }
             }
+            let mut recycled_files = self.recycled_files.write();
+            // If we could recycle the file, we would put the file_id into the
+            // `RecycleFileCollections` for the future reusing.
+            if recycled_files.valid_to_push(1) {
+                // A stale file is purged, so we should update the capacity of
+                // `ReycleFileCollection` by "current_capacity + 1".
+                recycled_files.upd_capacity(1);
+                recycled_files.push_one_file(file_id);
+                continue;
+            } else {
+                // If we could not push the stale file into the RecycleFileCollections, it means
+                // that:
+                // [1]: the RecycleFileCollections is invalid for inserting, that is, it's not
+                //      set by configuration.
+                // [2]: the RecycleFileCollections has no extra space for the staled file.
+                let first_stale_fid = recycled_files.pop_one_file();
+                match first_stale_fid {
+                    None => {
+                        // Invalid RecycleFileCollections
+                        // do nothing.
+                    }
+                    Some(fid) => {
+                        // Delete the first stale file in RecycleFileCollections to free space
```

Comment:
No need. You can just delete the new file here. There's no significant difference between an old file and a newly purged one.

Reply:
Oh, i should make an extra annotation here.
The reason why i did the implementations as this was that i wanted to keep the durable files(with FileSeq) consistent without holes. Because the files in the `RecycleFileCollection`had not be `renamed` with new `FileId` until the `pipe` need a new file for writting. And if we just delete the new file here, there might exist a circumstance that the `engine` was crashed unexpectedly after the deletion. Then, the durable `files` would had a `hole` , which would break the consistent of file sequences. It would cause that the engine could not recover to the expected state as we expected.

Reply:
Make sense. Then I suggest reusing the files collection, but using a counter `obsolete_files` to denote files that can be recycled:
```
fn purge_to(n):
  self.obsolete_files = n;
  if self.files.len() > max_file_count { /* physically delete some files; */ }

fn rotate:
  if self.obsolete_files > 0 {
    let f = self.files.pop_back();
  } else {
    // create
  }
```

Reply:
Got it, i'll modify it and redesign the related implementation on RecycleFileCollection.